### ISSUE TITLE
Customer property setter Map of Schema always instance of Document

### ISSUE THREAD

**Issue by miniplus** [Automattic/mongoose#7272]
**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**
Consider having 2 schemas (Parent and Child).
The Parent has a Map of ChildSchema.

The ChildSchema setters' context `this` is always an instance of _Document_, even when a model.find() was done. It is never an instance of _Query_ (unlike) the ParentSchema.

As far as I can tell this only happens when the ChildSchema is being used as a Map. When the ChildSchema is used directly as a Subdocument or an array of Subdocuments (not a Map), this context is correct (this might be false, I'm not 100% sure but in my tests it seemed to be ok).

https://mongoosejs.com/docs/api.html#schematype_SchemaType-set refers to this at the end of the section (starting with _You can also use setters to modify other properties on the document..._)

**If the current behavior is a bug, please provide the steps to reproduce.**
```js
export const ParentSchema: mongoose.Schema = new mongoose.Schema(
  {
   name: String,
   children: {
      type: Map,
      of: ChildSchema,
    }
  }
);

export const ChildSchema: mongoose.Schema = new mongoose.Schema(
  {
   age: {
      type: Number,
      set: setName,
    },
   yearBorn: Number,
   name: String
  }
);

function setAge(age: Number) {
  // Never a Query
  if (this instanceof mongoose.Query) {
     return age;
  }

  // Always a Document
  if (this instanceof mongoose.Document) {
      this.yearBorn = 2018;
      return age;
  }
}
```
**What is the expected behavior?**
`this` is _always_ an instance of Document, causing the logic inside to be run on model.find() (in this case, setting yearBorn to 2018. Just like the parent Schema, this should be an instance of Query in that case.

**Please mention your node.js, mongoose and MongoDB version.**
Node: v8.11.4
mongoose: v5.3.13
MongoDB v4.0.3

**Comment by vkarpov15** [comment]
Can you please clarify how you're getting the setter to run with incorrect context? The below script fails, but with a different bug than what you specified:

```javascript
const assert = require('assert');
const mongoose = require('mongoose');
mongoose.set('debug', true);

const GITHUB_ISSUE = `gh7272`;
const connectionString = `mongodb://localhost:27017/${ GITHUB_ISSUE }`;
const { Schema } = mongoose;

run().then(() => console.log('done')).catch(error => console.error(error.stack));

async function run() {
  await mongoose.connect(connectionString);
  await mongoose.connection.dropDatabase();

  const ChildSchema = new mongoose.Schema({
    age: {
      type: Number,
      set: setAge,
    },
    yearBorn: Number,
    name: String
  });

  const ParentSchema = new mongoose.Schema({
    name: String,
    children: {
      type: Map,
      of: ChildSchema,
    }
  });

  function setAge(age) {
    console.log('Setter', this);
    // Never a Query
    if (this instanceof mongoose.Query) {
      return age;
    }

    // Always a Document
    if (this instanceof mongoose.Document) {
      this.yearBorn = 2018;
      return age;
    }
  }

  const Parent = mongoose.model('Parent', ParentSchema);

  await Parent.find({ 'children.age': 21 });
}
```

Gives me an error `CastError: Cast to Embedded failed for value "21" at path "children.$*" for model "Parent"`

**Comment by miniplus** [comment]
@vkarpov15 In your reproduction, it makes sense that you're receiving an error as `children.age` doesn't exists.
`children` is a **map of ChildSchema**, meaning it'll have the following structure:

```js
children: {
    kenny: {
        age: 21
    }
}
```

So in order to query correctly it's supposed to be `.find({'children.kenny.age': 21})`.
**You might be on to this being a separate bug however?**

---

Back to the issue I was referring to (thanks for the reproduction snippet btw):
```js
const mongoose = require('mongoose');
mongoose.set('debug', true);

const GITHUB_ISSUE = `gh7272`;
const connectionString = `mongodb://localhost:27017/${ GITHUB_ISSUE }`;
const { Schema } = mongoose;

run().then(() => console.log('done')).catch(error => console.error(error.stack));

async function run() {
  await mongoose.connect(connectionString);
  await mongoose.connection.dropDatabase();

  const ChildSchema = new mongoose.Schema({
    age: {
      type: Number,
      set: setAge,
    },
    yearBorn: Number,
    name: String
  });

  const ParentSchema = new mongoose.Schema({
    name: String,
    children: {
      type: Map,
      of: ChildSchema,
    }
  });

  function setAge(age) {
    console.log('Setter', this);
    if (this instanceof mongoose.Query) {
      console.log('this is Query');
      return age;
    }

    if (this instanceof mongoose.Document) {
      console.log('this is Document and yearBorn is set as well');
      this.yearBorn = 2018;
      return age;
    }
  }

  const Parent = mongoose.model('Parent', ParentSchema);
  const parent = new Parent();
  
  parent.set('children.kenny', {
    age: 44
  });

  await parent.save();
  
  console.log('---Now Querying---');
  await Parent.find({'children.kenny.age': 44});
}
```

When saving the parent, the `setAge` setter is called correctly and the `this` context is `Document`.
When later we query, the setter is called as well (as it should according to Mongoose docs), but the setter is called **twice** so it seems (unlike my initial post, stating `this` is never `Query`).
The first time the setter is called, the `this` context is `Query` (correct).
But the second time it's called, the `this` context is `Document`, causing my setter logic (setting yearBorn) to also run on a Query.

**Comment by vkarpov15** [comment]
You're right, my original repro script threw the correct error. `children.age` should be a document, not a number.

The issue was that Mongoose was calling the setter when initializing the subdoc. Fix will be in 5.3.15 :+1:
